purchases/src/main/kotlin/com/revenuecat/purchases/google/BillingWrapper.kt

```diff
@@ -119,7 +128,22 @@ internal class BillingWrapper(
             billingClient?.let {
                 if (!it.isReady) {
                     log(LogIntent.DEBUG, BillingStrings.BILLING_CLIENT_STARTING.format(it))
-                    it.startConnection(this)
+                    if (startConnectionAttempts < MAX_RECONNECTION_ATTEMPTS) {
+                        startConnectionAttempts++
+                        try {
+                            it.startConnection(this)
+                        } catch (e: IllegalStateException) {
+                            log(
+                                LogIntent.GOOGLE_ERROR,
+                                BillingStrings.ILLEGAL_STATE_EXCEPTION_WHEN_CONNECTING.format(e),
+                            )
+                        }
+                    } else {
+                        // There's an IllegalStateException on Samsung devices that try to bind to the service more
+                        // than 999 times. This is a workaround to avoid that.
+                        log(LogIntent.DEBUG, BillingStrings.BILLING_CLIENT_TOO_MANY_RETRIES)
+                        it.endConnection()
```

Comment:
I am not sure what to do with the counter in this case. Restarting it here would make it useless. Ideally we would restart it when it finishes disconnecting but `onBillingServiceDisconnected` doesn't get triggered after `endConnection` finishes. Also... if we are trying to connect, should we even trying to call `endConnection`?

Maybe we should just not do anything here

Reply:
Hmm so there is no way of knowing when `endConnection` finishes? That's annoying... In this case, I would:
- Allow to keep retrying, as long as it doesn't crash (with the try-catch). I don't think we should stop retrying without understanding whether this issue is temporary or permanent.
- Propagate an error to anything in the `serviceRequests` queue so anything waiting for the connection gets notified on the error (and ideally propagate it to the user).

Reply:
Propagating is a good idea, the only thing is that how would the developer react to the error? There will be one point where all calls to our SDK that use the BillingWrapper will immediately return an error right?

Reply:
Well the developer should react the same way as any other billing connection error I guess... which probably would be showing an error whenever trying to get to the paywall. 

And well, that about immediately, I guess once it reaches that 999 threshold, then yeah, calls would probably immediately fail

Reply:
+1 on propagating, I don't think silent-failing is a better outcome. Even if a developer decides to not catch it and hard crash, that might be a better outcome in that presumably on reopen connection should work